File: miniML/examples/lazy.ml
let r : ref int = ref 1 in
let x : lazy int = lazy (let u : () = r := !r + 1 in !r) in
let f : lazy int -> int = fun (z : lazy int) -> force z * (force z + 1) in
let y : lazy int = lazy (f x) in
let w : lazy () = lazy (let u : () = r := 1234 in ()) in
!r * f y

File: miniML/src/MiniML/Lazy.hs
module MiniML.Lazy where
{- 
Student name: 
Student ID:
-}

import MiniML.Syntax

implementLazy :: Exp -> Exp
implementLazy (Prim p "lazy" e) = error "TODO!"
implementLazy (Prim p "force" e) = error "TODO!"
implementLazy _ = error "TODO!"

{- Question1:  How would you define a stream (lazy list)?
Answer: 

Question 2: How would you implement a lazy computation of the n-th Fibonacci number?
Answer: 

-}